### PR TITLE
523 add date time picker

### LINKED ISSUE #523: Add a date time picker for cross browser compatibility and force 15 min increments
### User story
Add a date time picker for cross browser compatibility and force 15 min increments

### PR DESCRIPTION
**Pull request by a-drew**
##### _What existing problem does the pull request solve?_

Adding new DateTimePicker to replace all existing inputs of `datetime-local` type. This will help us avoid browser dependant issues and be more robust during dusk testing (local vs CI). Additionally, this package allows us to easily manage intervals and force 15m timeslots during booking creation.

Screen Shot 2021-04-03 at 10 51 35 PM

##### _Demonstrate the code is solid. </br> Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI._

- tests pass, sonar is clowning hard and ignoring valid tests

##### _List all issues this PR resolves_
- closes #523

##### _Complete these items before requesting reviews_
- [x] Static Analysis Rules Validated
- [x] All commits follow naming conventions

### Sub-tasks Checklist
- [x] UI Implementation
- [x] UI Test
- [x] Backend Implementation
- [x] Backend Tests
- [ ] Integration Tests
- [ ] System Tests
- [ ] Documentation

### CODE REVIEW

File: resources/js/app.js [hunk 1]
```diff
@@ -7,14 +7,16 @@ import { InertiaApp } from '@inertiajs/inertia-vue';
 import { InertiaForm } from 'laravel-jetstream';
 import PortalVue from 'portal-vue';
 import Calendar from 'v-calendar/lib/components/calendar.umd'
-import DatePicker from 'v-calendar/lib/components/date-picker.umd'
 import DateFormatter from "@src/Plugins/date-formatter";
 
+import VueCtkDateTimePicker from 'vue-ctk-date-time-picker/dist/vue-ctk-date-time-picker.umd';
+import 'vue-ctk-date-time-picker/dist/vue-ctk-date-time-picker.css';
```
Comment (alexstojda) [inline resources/js/app.js:13]: Including CSS like this will work? TIL

**Review by alexstojda — DISMISSED** [review]
Code looks good, nice work. 

Sonar seems bugged as well, so I'll sign off on an merge with admin privileges to override the sonar failure

File: resources/js/app.js (hunk 1 above)
Comment (a-drew) [reply resources/js/app.js:13]: Ye, definitely odd. comes from install instructions: https://github.com/chronotruck/vue-ctk-date-time-picker

Also if I don't add this here but try to import in the DateTimePicker.vue file, Vue will throw warnings about infinite loops and the inputs won't work 🤷

Comment (alexstojda) [reply resources/js/app.js:13]: Interesting, and if you include it in app.css instead of here it also throws warnings?

**Comment by a-drew** [comment]
> Code looks good, nice work.
> 
> Sonar seems bugged as well, so I'll sign off on an merge with admin privileges to override the sonar failure

Can confirm results of https://sonarcloud.io/component_measures?id=CSU-Booking-Platform_application&metric=new_coverage&pullRequest=600&selected=CSU-Booking-Platform_application%3Aapp%2FHttp%2FControllers%2FBookingRequestController.php&view=list are bugged.

`BookingRequestController` is covered extensively by `BookingRequestControllerTest`, `FilterPersonalBookingRequestsTest`, `FilterBookingRequestsTest`.

File: resources/js/app.js (hunk 1 above)
Comment (a-drew) [reply resources/js/app.js:13]: nah css should be fine. It's the component import that do the infinite loop thing. something weird with UMD modules, ES6 all the way

**Review by danseremet — CHANGES_REQUESTED** [review]
# Review
I've tested locally the changes you made for the date picker.

# Result
- In the create a booking modal:
  - The date picker works correctly although the containing modal is small.
- In the settings for civic holidays it works correctly.

# Issues with sonar:
I've ran the tests locally and checked the coverage.
I get the same issues as sonar cloud. The BookingRequestController is not being hit by some of the tests covering the 'index', 'createInit' and 'create' functions.

I've tried debugging the tests inside BookingRequestControllerTest, and the tests that trigger the covered functions work, but those that I expect to hit the above-mentioned functions do not.
I've tried replacing your direct $this-get to use $this->actingAs and I was able to hit breakpoints for the functions not covered. If you were to replace it with that, I believe the functions will be covered and sonar will stop complaining.


## Example

image

**Note: I think the asserts are incorrect for some of the tests after I tried that change, I'll let you find why :smile:**

**Comment by jonathanp16** [comment]
Looks very nice and will definitely be sprinkling this everywhere, just need some tests to bump coverage.

**Review by alecadub — APPROVED** [review]
The code looks good ! Once you bump the coverage, 🚢 🇮🇹

**Comment by a-drew** [comment]
<a href="https://imgflip.com/i/54cvwt"></a>

TLDR: I can't bump coverage for something that's already covered. sonar is wrong, fight me. 

##### Actual arguments and logic below.

### Coverage for unrelated BookingController tests

> I get the same issues as sonar cloud. The BookingRequestController is not being hit by some of the tests covering the 'index', 'createInit' and 'create' functions.

> I've tried debugging the tests inside BookingRequestControllerTest, and the tests that trigger the covered functions work, but those that I expect to hit the above-mentioned functions do not.
> I've tried replacing your direct $this-get to use $this->actingAs and I was able to hit breakpoints for the functions not covered. If you were to replace it with that, I believe the functions will be covered and sonar will stop complaining.

Appreciate the thorough look and I do agree that the create / createInit functions are not being covered appropriately since our introduction of createInit. However modifying tests for those endpoints will not improve sonar's opinion about this PR, they aren't related to my changes in the controller. Besides, these tests are still passing or we wouldn't even reach sonar. 

### Sonar is wack

My particular gripe with sonar is the erroneous interpretation of the filter / filterUserBooking coverage. As far as I am concerned, the only missing coverage I am responsible for in this PR are the very few line changes I made to those two method. These methods are covered by Evan's unit tests `FilterPersonalBookingRequestsTest`, `FilterBookingRequestsTest`. Those tests are build properly, authenticated properly and hit the endpoints as expected.

Here's my local coverage results ( I encourage anyone that reviews to run a local coverage scan as well ) :

Screen Shot 2021-04-04 at 12 40 27 PM

Screen Shot 2021-04-04 at 12 40 36 PM

Notice the lines I modified yesterday are fully covered. 

Meanwhile Sonar is outright saying that those methods are not covered at all : 
Screen Shot 2021-04-04 at 12 36 12 PM

Screen Shot 2021-04-04 at 12 36 28 PM

So as I originally stated, it is impossible for me to bump coverage for something that's already covered.

_ty for coming to my ted talk_

**Review by danseremet — APPROVED** [review]
😂

**Review by jonathanp16 — APPROVED** [review]
The evidence is there
